### PR TITLE
Support dynamic RAG syntax through the gRPC API

### PR DESCRIPTION
**Pull request by tsmith023**
### What's being changed:

- Adds new messages allowing for input of provider-specific config at query time
- Adds new messages allowing for output of provider-specific metadata at query time
- Adds acceptance tests for dynamic generative searches through gRPC
- Fixes aspects of the module implementation
- Attempts to clean tech debt from the gRPC API by modularising code
- Ensures the minimal amount of implicit behaviour between the module and gRPC layers

### What is still to come in other PRs:

- Modifying the more complex generative modules (`aws`, `openai`, `palm`/`google`) to allow for query-time definition of all available configuration options
- Define new module-specific namespace structs to contain both the `classSettings` vars and the `Params` vars in a single location
- Removing the `EXPERIMENTAL_` env var such that the feature is generally available

### Review checklist

- [x] Documentation has been updated, if necessary. Link to changed documentation:
- [x] Chaos pipeline run or not necessary. Link to pipeline:
- [x] All new code is covered by tests where it is reasonable.
- [x] Performance tests have been run or not necessary.

### CODE REVIEW

File: grpc/proto/v1/generative.proto [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+    GenerativeAnyscale anyscale = 3;
+    GenerativeAWS aws = 4;
+    GenerativeCohere cohere = 5;
+    GenerativeDummy dummy = 6;
+    GenerativeMistral mistral = 7;
+    GenerativeOctoAI octoai = 8;
+    GenerativeOllama ollama = 9;
+    GenerativeOpenAI openai = 10;
+    GenerativeGoogle google = 11;
+  }
```
Comment (antas-marcin) [inline grpc/proto/v1/generative.proto:46]: there's also a `generative-databricks` module, but we can add support for it in a separate PR

File: test/modules/many-generative/many_generative_test.go [hunk 2]
```diff
@@ -85,8 +85,8 @@ func testGenerativeManyModules(host, ollamaApiEndpoint, region, gcpProject strin
 				params: `aws:{temperature:0.9 model:"ai21.j2-mid-v1"}`,
 			},
 			{
-				name:   "google",
-				params: `google:{topP:0.8 topK:30 model:"chat-bison"}`,
+				name:   "palm",
```
Comment (antas-marcin) [inline test/modules/many-generative/many_generative_test.go:88]: it suppose to be google, why did you change that?

File: grpc/proto/v1/generative.proto [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+    GenerativeAnthropicMetadata anthropic = 1;
+    GenerativeAnyscaleMetadata anyscale = 2;
+    GenerativeAWSMetadata aws = 3;
+    GenerativeCohereMetadata cohere = 4;
+    GenerativeDummyMetadata dummy = 5;
+    GenerativeMistralMetadata mistral = 6;
+    GenerativeOctoAIMetadata octoai = 7;
+    GenerativeOllamaMetadata ollama = 8;
+    GenerativeOpenAIMetadata openai = 9;
+    GenerativePaLMMetadata palm = 10;
```
Comment (antas-marcin) [inline grpc/proto/v1/generative.proto:233]: let's use new naming instead of `PaLM` let's use `Google`

File: test/modules/generative-palm/generative_palm_test.go [hunk 4]
```diff
@@ -121,9 +124,26 @@ func testGenerativePaLM(host, gcpProject string) func(t *testing.T) {
 					planets.CreateTweetTest(t, class.Class)
 				})
 				t.Run("create a tweet with params", func(t *testing.T) {
-					params := "google:{topP:0.1 topK:40}"
+					params := "palm:{topP:0.1 topK:40}"
```
Comment (antas-marcin) [inline test/modules/generative-palm/generative_palm_test.go:127]: I can see still here that you are using `palm` namespace instead of `google`

Comment (tsmith023) [reply test/modules/generative-palm/generative_palm_test.go:127]: Thank you for spotting it!

File: modules/generative-palm/clients/palm.go [hunk 5]
```diff
@@ -344,7 +344,21 @@ func (v *palm) parseResponse(statusCode int, bodyBytes []byte, debug *modulecapa
 }
 
 func (v *palm) getResponseParams(params map[string]interface{}) map[string]interface{} {
-	return map[string]interface{}{"google": params}
+	return map[string]interface{}{googleparams.Name: params}
+}
+
+func GetResponseParams(result map[string]interface{}) *responseParams {
```
Comment (antas-marcin) [inline modules/generative-palm/clients/palm.go:350]: does this method need to be public?

Comment (tsmith023) [reply modules/generative-palm/clients/palm.go:350]: I have those methods as public so that the gRPC package can extract the correct type from the `map[string]any` that it is working with, which comes from the `traverser.GetClass` method at the root of the `Search` method

My thought was to keep `responseParams` private and have the public method allowing for this requirement. The alternative I saw was to have `ResponseParams` in this package and then `getGoogleResponseParams` in the gRPC package but in my view this would lead to code in different packages depending on the specific format of the `result map[string]any` variable

In this way, the definition of the returned `map[string]any` var, which happens on L347 here can coexist next to the function that maps it back into the `responseParams` type from which it was originally mapped

Hopefully that makes sense 😅
